hoodie-client/src/main/java/com/uber/hoodie/index/bloom/HoodieBloomIndex.java

```diff
@@ -0,0 +1,380 @@
+            }
+            logger.info("TotalRecords: " + totalRecords + ", TotalFiles: " + totalFiles + ", TotalAffectedPartitions:" + recordsPerPartition.size());
+        }
+
+        // each partition will have an item per comparison.
+        int parallelism = (int) (totalComparisons/ MAX_ITEMS_PER_SHUFFLE_PARTITION + 1);
+        logger.info("Auto computed parallelism :" + parallelism + ", totalComparisons: " + totalComparisons);
+        return parallelism;
+    }
+
+    /**
+     * Its crucial to pick the right parallelism.
+     *
+     * totalSubPartitions : this is deemed safe limit, to be nice with Spark.
+     * inputParallelism : typically number of input file splits
+     *
+     * We pick the max such that, we are always safe, but go higher if say a there are a lot of
+     * input files. (otherwise, we will fallback to number of partitions in input and end up with
+     * slow performance)
+     */
+    private int determineParallelism(int inputParallelism, int totalSubPartitions) {
+        // If bloom index parallelism is set, use it to to check against the input parallelism and take the max
+        int indexParallelism = Math.max(inputParallelism, config.getBloomIndexParallelism());
+        int joinParallelism = Math.max(totalSubPartitions, indexParallelism);
+        logger.info("InputParallelism: ${" + inputParallelism + "}, " +
+                "IndexParallelism: ${" + config.getBloomIndexParallelism() + "}, " +
+                "TotalSubParts: ${" + totalSubPartitions + "}, " +
+                "Join Parallelism set to : " + joinParallelism);
+        return joinParallelism;
+    }
+
+    /**
+     * Load all involved files as <Partition, filename> pair RDD.
+     */
+    @VisibleForTesting
+    List<Tuple2<String, BloomIndexFileInfo>> loadInvolvedFiles(List<String> partitions, final HoodieTable<T> hoodieTable) {
+        // Obtain the latest data files from all the partitions.
+        List<Tuple2<String, HoodieDataFile>> dataFilesList = jsc.parallelize(partitions, Math.max(partitions.size(), 1))
+                .flatMapToPair(partitionPath -> {
+                    java.util.Optional<HoodieInstant> latestCommitTime =
+                            hoodieTable.getCommitTimeline().filterCompletedInstants().lastInstant();
+                    List<Tuple2<String, HoodieDataFile>> filteredFiles = new ArrayList<>();
+                    if (latestCommitTime.isPresent()) {
+                         filteredFiles =
+                                hoodieTable.getROFileSystemView().getLatestDataFilesBeforeOrOn(partitionPath,
+                                        latestCommitTime.get().getTimestamp())
+                                        .map(f -> new Tuple2<>(partitionPath, f))
+                                        .collect(toList());
+                    }
+                    return filteredFiles.iterator();
+                }).collect();
+
+        if (config.getBloomIndexPruneByRanges()) {
```

Comment:
is there a way to avoid multiple if-else's ? I guess using some sort of Handler is an overkill but probably a way to read minMax keys in both paths and mark minRecordKey and maxRecordKey as optional if they are not present (since we will be reading the footer in both cases anyways ?). Just a thought.